### PR TITLE
Automatically resolve failed assembly resolves to alongside test library

### PR DESCRIPTION
**Pull request by davkean**
Fixes: #9490.

Resolves failed assembly resolves to assemblies that live alongside the dll that contains this test framework.

To avoid running into the.NET EventListener bug https://github.com/dotnet/roslyn/issues/6358, we turn off AppDomain isolation in xUnit.Unfortunately, this changes binding behavior in two ways;
 
1) AppBase becomes the location of xunit.console.[arch].exe executable, instead of the test library.

2) Binding redirect and other binding settings are not respected because AppDomain 0 uses whatever are in xunit.console.[arch].exe.config.

This is problematic for a variety of tests that expect this to work. For example, the project system    tests have an indirect reference on System.Collections.Immutable, 1.1.36 however, we only deploy    System.Collections.Immutable, 1.1.37, which, without a binding redirects, causes a version mismatch on .NET Framework.

This attempts to solve this by wrapping xunit's test framework to automatically hook onto AppDomain.AssemblyResolve for failed resolves and resolve to dlls locally. This is currently,
hardened to only work for project system tests, but we should be able to long term make this work for all tests that need special binding or binding redirects to be in play even though they disable AppDomain in xunit.

### CODE REVIEW

**Comment by davkean** [comment]
tag @dotnet/roslyn-infrastructure @srivatsn @dotnet/roslyn-ide @amcasey @ManishJayaswal

File: src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ResolveDependenciesLocallyTestFramework.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```
Comment (davkean) [inline src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ResolveDependenciesLocallyTestFramework.cs]: Interesting code starts here

File: src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ResolveDependenciesLocallyTestFramework.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+using System;
+using System.IO;
+using System.Reflection;
+using Xunit;
+using Xunit.Abstractions;
+using Xunit.Sdk;
+
+namespace Microsoft.VisualStudio
+{
+    // Resolves failed assembly resolves to assemblies that live alongside the dll that contains this test
```
Comment (amcasey) [inline src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ResolveDependenciesLocallyTestFramework.cs]: Consider making this a doc comment.

File: src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ResolveDependenciesLocallyTestFramework.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        }
+
+        public ITestFrameworkExecutor GetExecutor(AssemblyName assemblyName)
+        {
+            return _underlyingFramework.GetExecutor(assemblyName);
+        }
+
+        public void Dispose()
+        {
+            AppDomain.CurrentDomain.AssemblyResolve -= OnAssemblyResolve;
```
Comment (amcasey) [inline src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ResolveDependenciesLocallyTestFramework.cs]: This is a no-op the second time it's called?

File: src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ResolveDependenciesLocallyTestFramework.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            _underlyingFramework.Dispose();
+        }
+
+        private Assembly OnAssemblyResolve(object sender, ResolveEventArgs args)
+        {
+            string directoryPath = Path.GetDirectoryName(GetType().Assembly.Location);
+
+            AssemblyName name = new AssemblyName(args.Name);
+
+            return Assembly.LoadFrom(Path.Combine(directoryPath, name.Name + ".dll"));
```
Comment (amcasey) [inline src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ResolveDependenciesLocallyTestFramework.cs]: Would there be any value in searching for ".exe" if this fails?

**Comment by amcasey** [comment]
I know little about this, but your explanation and fix seem plausible.

File: src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ResolveDependenciesLocallyTestFramework.cs (hunk 3 above)
Comment (davkean) [inline src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ResolveDependenciesLocallyTestFramework.cs]: Yes.

File: src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ResolveDependenciesLocallyTestFramework.cs (hunk 4 above)
Comment (davkean) [inline src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ResolveDependenciesLocallyTestFramework.cs]: Not right now, not expecting or seeing any tests that have dependencies on exes.

File: src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ResolveDependenciesLocallyTestFramework.cs (hunk 2 above)
Comment (davkean) [inline src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ResolveDependenciesLocallyTestFramework.cs]: Looking at moving this to a more central place right now for another change - will do it that then.

File: src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ResolveDependenciesLocallyTestFramework.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    //    settings are in xunit.console.[arch].exe.config.
+    //
+    // This is problematic for a variety of tests that expect this to work. For example, the project system 
+    // tests have an indirect reference on System.Collections.Immutable, 1.1.36 however, we only deploy 
+    // System.Collections.Immutable, 1.1.37, which, without a binding redirects, causes a version mismatch on
+    // .NET Framework.
+    //
+    // The aim of this class is avoid the need of binding redirects, by always resolving to whatever assembly 
+    // is deployed alongside the test framework, we avoid the need
+    public class ResolveDependenciesLocallyTestFramework : ITestFramework
```
Comment (davidfowl) [inline src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ResolveDependenciesLocallyTestFramework.cs]: Do you guys have any native binaries that get deployed to the output folder? That's one of the things that breaks with this approach.

Comment (davkean) [inline src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ResolveDependenciesLocallyTestFramework.cs]: No we don't, but if we did you could use DefaultDllImportSearchPathsAttribute to influence this to look alongside the assembly.

**Pull request merged**